topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,623 @@
+		Result:           &result,
+	}
+	if decoder, err = mapstructure.NewDecoder(config); err != nil {
+		return nil, err
+	}
+	if err := decoder.Decode(deviceMap["blockdevices"]); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (p *ProbeHandler) unregisterBlockdev(id string) {
+	p.Lock()
+	defer p.Unlock()
+
+	info, ok := p.blockdevMap[id]
+	if !ok {
+		return
+	}
+
+	p.Ctx.Graph.Lock()
+	if err := p.Ctx.Graph.DelNode(info.Node); err != nil {
+		p.Ctx.Graph.Unlock()
+		p.Ctx.Logger.Error(err)
+		return
+	}
+	p.Ctx.Graph.Unlock()
+
+	delete(p.blockdevMap, id)
+}
+
+func (p *ProbeHandler) connect() error {
+	var (
+		cmdOut []byte
+		err    error
+		intf   interface{}
+		result Devices
+	)
+
+	lsblkPath := p.Ctx.Config.GetString("agent.topology.blockdev.lsblk_path")
+
+	if lsblkPath == "" {
+		lsblkPath = "/usr/bin/lsblk"
+	}
+
+	if cmdOut, err = exec.Command(lsblkPath, "-pO", "--json").Output(); err != nil {
+		return err
+	}
+
+	if err = json.Unmarshal([]byte(cmdOut[:]), &intf); err != nil {
+		return err
+	}
+
+	if err = mapstructure.WeakDecode(intf, &result); err != nil {
+		return err
+	}
+
+	// loop through the devices in the current map to make sure they haven't
+	// been removed
+	for _, current := range p.blockdevMap {
+		p.deleteIfRemoved(current, result.Blockdevices)
+	}
+
+	for i := range result.Blockdevices {
+		p.registerBlockdev(result.Blockdevices[i], "")
+	}
+
+	return nil
+}
+
+func (p *ProbeHandler) getID(blockdev BlockDevice) string {
+	if blockdev.WWN != "" {
+		return blockdev.WWN
+	}
+	if blockdev.Serial != "" {
+		return blockdev.Serial
+	}
+	return blockdev.Name
+}
+
+func (p *ProbeHandler) getName(blockdev BlockDevice) string {
+	if blockdev.Mountpoint != "" {
+		return blockdev.Mountpoint
+	}
+	return blockdev.Name
+}
+
+func (p *ProbeHandler) newMetricsFromBlockdev(blockdevPath string) *BlockMetric {
+	var (
+		cmdOut []byte
+		err    error
+		stats  sysstat
+		intf   interface{}
+	)
+
+	iostatPath := p.Ctx.Config.GetString("agent.topology.blockdev.iostat_path")
+
+	if iostatPath == "" {
+		iostatPath = "/usr/bin/iostat"
+	}
+
+	if cmdOut, err = exec.Command(iostatPath, "-dx", "-o", "JSON", blockdevPath).Output(); err != nil {
+		return nil
+	}
+
+	if err = json.Unmarshal([]byte(cmdOut[:]), &intf); err != nil {
+		return nil
+	}
+
+	if err = mapstructure.WeakDecode(intf, &stats); err != nil {
+		return nil
+	}
+
+	if stats.Sysstat.Hosts == nil || len(stats.Sysstat.Hosts) == 0 || stats.Sysstat.Hosts[0].Statistics == nil ||
+		len(stats.Sysstat.Hosts[0].Statistics) == 0 || stats.Sysstat.Hosts[0].Statistics[0].Metrics == nil ||
+		len(stats.Sysstat.Hosts[0].Statistics[0].Metrics) == 0 {
+		return nil
+	}
+	return stats.Sysstat.Hosts[0].Statistics[0].Metrics[0].MakeCopy()
+}
+
+func (p *ProbeHandler) updateBlockDevMetric(now, last time.Time) {
+	for _, blockdev := range p.blockdevMap {
+		node := blockdev.Node
+		if node == nil {
+			continue
+		}
+		path, err := node.GetField("Path")
```

Comment:
This should be done under graph lock.

nit: you can probably use `GetFieldString` (and therefore "%s" in the Sprintf below).